otp/pool/solution/poolie_server.erl

```diff
@@ -0,0 +1,92 @@
+-module(poolie_server).
+
+-behaviour(gen_server).
+
+-export([start/2, stop/0, start_link/2, run/3, run/2, pool_info/0]).
+-export([init/1, handle_call/3, handle_cast/2, handle_info/2, terminate/2, code_change/3]).
+
+-record(state, {limit, sup, idle=gb_sets:empty(), busy=gb_sets:empty()}).
+-define(SERVER, ?MODULE).
+
+%% API
+
+run(M, F, A) when is_list(A) ->
+  Msg = gen_server:call(?SERVER, {work, {M, F, A}}),
+  io:format(Msg).
+
+run(F, A) when is_function(F), is_list(A) ->
+  Msg = gen_server:call(?SERVER, {work, {F, A}}),
+  io:format(Msg).
+
+pool_info() ->
+  {PoolSize, Idle, Busy} = gen_server:call(?SERVER, info),
+  io:format("Pool has ~p workers.~nThere are ~p idle and ~p busy workers.~n", [PoolSize, Idle, Busy]).
+
+start(PoolSize, Sup) ->
+  gen_server:start({local, ?SERVER}, ?MODULE, {PoolSize, Sup}, []).
+
+start_link(PoolSize, Sup) ->
+  gen_server:start_link({local, ?SERVER}, ?MODULE, {PoolSize, Sup}, []).
+
+stop() ->
+  gen_server:call({local, ?SERVER}, stop).
+
+%% gen_server callbacks
+
+init({PoolSize, Sup}) ->
+  self() ! {start_worker_supervisor, Sup},
```

Comment:
Instead of passing around the supervisor to attach to yo should initiate it where it's expected (`poolie_sup`) and figure how to pass the workers list to this process afterwards.

Reply:
Refactored 4e3b3f1